WordPress/WordPressTest/PagesListTests.swift

```diff
@@ -0,0 +1,502 @@
+        let roundtrip: (String) throws -> Void = { string in
+            let pages = try Array<Page>(hierarchyListRepresentation: string, context: self.mainContext)
+            XCTAssertEqual(PageTree.hierarchyList(of: pages).hierarchyListRepresentation(), string)
+        }
+
+        try roundtrip("""
+            1
+              2
+            3
+              4
+                5
+              6
+            7
+            """)
+
+        try roundtrip("""
+            1
+              2
+                3
+              4
+                5
+              6
+            7
+            """)
+    }
+
+    private func parentPage(childrenCount: Int, additionalLevels: Int = 0) -> [Page] {
+        var pages = [Page]()
+
+        let parent = PageBuilder(mainContext).build()
+        parent.postID = NSNumber(value: randomID.next())
+        parent.parentID = 0
+        pages.append(parent)
+
+        let children = (0..<childrenCount).map { _ in
+            let child = PageBuilder(mainContext).build()
+            child.postID = NSNumber(value: randomID.next())
+            child.parentID = parent.postID
+            return child
+        }
+        pages.append(contentsOf: children)
+
+        if additionalLevels > 1 {
+            let nested = parentPage(childrenCount: childrenCount, additionalLevels: additionalLevels - 1)
+            nested[0].parentID = parent.postID
+            pages.append(contentsOf: nested)
+        }
+
+        return pages
+    }
+
+    private func makeAssertions(pages: [Page], file: StaticString = #file, line: UInt = #line) {
+        var start: CFAbsoluteTime
+
+        start = CFAbsoluteTimeGetCurrent()
+        let original = pages.hierarchySort()
+        NSLog("hierarchySort took \(String(format: "%.3f", (CFAbsoluteTimeGetCurrent() - start) * 1000)) millisecond to process \(pages.count) pages")
+
+        start = CFAbsoluteTimeGetCurrent()
+        let pageTree = PageTree()
+        pageTree.add(pages)
+        let new = pageTree.hierarchyList()
+        NSLog("PageTree took \(String(format: "%.3f", (CFAbsoluteTimeGetCurrent() - start) * 1000)) millisecond to process \(pages.count) pages")
+
+        start = CFAbsoluteTimeGetCurrent()
+        _ = pages.sorted { ($0.postID?.int64Value ?? 0) < ($1.postID?.int64Value ?? 0) }
+        NSLog("Array.sort took \(String(format: "%.3f", (CFAbsoluteTimeGetCurrent() - start) * 1000)) millisecond to process \(pages.count) pages")
+
+        let originalIDs = original.map { $0.postID! }
+        let newIDs = new.map { $0.postID! }
+        let diff = originalIDs.difference(from: newIDs).inferringMoves()
+        XCTAssertTrue(diff.count == 0, "Unexpected diff: \(diff)", file: file, line: line)
+    }
+}
+
+final class PageTree {
+
+    // A node in a tree, which of course is also a tree itself.
+    private class TreeNode {
+        var page: Page
+        var children = [TreeNode]()
+        var parentNode: TreeNode?
```

Comment:
Of these, the only one that mutates is `parentNode`. Should the other be constants? 

```suggestion
        let page: Page
        let children = [TreeNode]()
        var parentNode: TreeNode?
```

Reply:
Addressed in https://github.com/wordpress-mobile/WordPress-iOS/pull/22056/commits/fdd9e7267f1043927fc131edadfe3bc12126bccb. `page` is not mutable, but `children` is.